bcc.1
.\"
.\" Copyright (c) 2020-2025 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd March 8, 2025
.Dt BCC 1
.Os
.\" NAME
.Sh NAME
.Nm bcc
.Nd binary to C/C++ arrays converter
.\" SYNOPSIS
.Sh SYNOPSIS
.Nm bcc
.Op Fl 0cs
.Op Fl I Ar tab-num
.Op Fl i Ar space-num
.Op Fl t Ar type
.Ar filename
.Ar variable
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
utility converts the
.Ar input
file into a C or C++ array named
.Ar variable
that can be embedded as-is in the source code. A special
.Ar -
value can be passed as input which will read standard input instead.
.Pp
The
.Ar variable
can contain any character but only ones that are allowed in the C standard will
be kept, other will be replaced with a
.Dq _ .
Also, the extension (by finding the first .) is removed as well. This can be
handy when generating a lots of file based on their names during a build
process.
.Pp
Note: you must still not start a variable name with digits.
.Pp
The following options are available:
.Bl -tag -width space-num-xxx
.It Fl 0
Terminate the generated array with a trailing NUL.
.It Fl c
Generates a
.Em const
array.
.It Fl I Ar tab-num
Use
.Ar tab-num
count as leading tab indents.
.It Fl i Ar space-num
Use
.Ar space-num
count as leading space indents.
.It Fl s
Generates a
.Em static
array.
.It Fl t Ar type
Changes the array char type to store the data content which defaults to
.Ar unsigned char .
.El
.\" EXAMPLES
.Sh EXAMPLES
Create a static const array from an image.
.Bd -literal -offset indent
bcc -sc image.png image > image.h
.Ed
.Pp
Create a modifiable char NUL-terminated array from a text file
.Bd -literal -offset indent
bcc -0 -t char text.txt text > text.h
.Ed
.\" HISTORY
.Sh HISTORY
The
.Nm
tool is inspired by
.Nm xxd
utility but offers more flexibility over the the generated code.
.\" SEE ALSO
.Sh SEE ALSO
.Xr xxd 1
.Sh AUTHORS
The
.Nm
utility was written by
.An David Demelier Aq Mt markand@malikania.fr .


CHANGES.md
bcc CHANGES
===========

bcc 2.2.1 2025-03-08
--------------------

- Fix build on OpenBSD.

bcc 2.2.0 2023-08-19
--------------------

- Add `-t type` option to change default `unsigned char` data type.

bcc 2.1.0 2023-03-27
--------------------

- Add `pledge` support on OpenBSD.
- Remove `getopt()` usage for non POSIX support.

bcc 2.0.0 2021-04-13
--------------------

- Use `unsigned char` only.

bcc 1.0.0 2020-10-21
--------------------

- Initial 1.0.0 release.


bcc.c
/*
 * bcc.c -- binary to C/C++ arrays converter
 *
 * Copyright (c) 2020-2025 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <errno.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#if defined(__OpenBSD__)
#       include <unistd.h>
#endif

#include "arg.h"

static const char *charset = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_";
static const char *ftype = "unsigned char";
static char findentchar = '\t';
static int findent = 1, fconst, fnull, fstatic;

static void
usage(void)
{
	fprintf(stderr, "usage: bcc [-0cs] [-I tab-num] [-i space-num] [-t type] input variable\n");
	exit(1);
}

static void
die(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	fputs("abort: ", stderr);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	exit(1);
}

static char *
mangle(char *variable)
{
	char *p;
	size_t pos;

	/* Remove extension. */
	if ((p = strrchr(variable, '.')))
		*p = '\0';

	/* Remove disallowed characters. */
	while ((pos = strspn(variable, charset)) != strlen(variable))
		variable[pos] = '_';

	return variable;
}

static void
indent(void)
{
	for (int i = 0; i < findent; ++i)
		putchar(findentchar);
}

static void
put(int ch)
{
	printf("0x%02hhx", (unsigned char)ch);
}

static void
process(const char *input, const char *variable)
{
	FILE *fp;
	int ch, col = 0;

	if (strcmp(input, "-") == 0)
		fp = stdin;
	else if (!(fp = fopen(input, "rb")))
		die("%s: %s\n", input, strerror(errno));

	if (fstatic)
		printf("static ");
	if (fconst)
		printf("const ");

	printf("%s %s[] = {\n", ftype, variable);

	for (ch = fgetc(fp); ch != EOF; ) {
		if (col == 0)
			indent();

		put(ch);

		if ((ch = fgetc(fp)) != EOF || fnull)
			printf(",%s", col < 3 ? " " : "");

		if (++col == 4) {
			col = 0;
			putchar('\n');
		}

		/* Add final '\0' if required. */
		if (ch == EOF && fnull) {
			if (col++ == 0)
				indent();

			put(0);
		}
	}

	if (col != 0)
		printf("\n");

	puts("};");
	fclose(fp);
}

int
main(int argc, char *argv[])
{
#if defined(__OpenBSD__)
	if (pledge("rpath stdio", NULL) < 0)
		die("abort: %s\n", strerror(errno));
#endif

	ARGBEGIN {
	case '0':
		fnull = 1;
		break;
	case 'c':
		fconst = 1;
		break;
	case 'I':
		findentchar = '\t';
		findent = atoi(EARGF(usage()));
		break;
	case 'i':
		findentchar = ' ';
		findent = atoi(EARGF(usage()));
		break;
	case 's':
		fstatic = 1;
		break;
	case 't':
		ftype = EARGF(usage());
		break;
	default:
		usage();
		break;
	} ARGEND

	if (argc < 2)
		usage();

	process(argv[0], mangle(argv[1]));
}
